cdap-ui/app/cdap/components/SpotlightSearch/index.js

```diff
@@ -0,0 +1,245 @@
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import React, {Component} from 'react';
+import classnames from 'classnames';
+import {MySearchApi} from 'api/search';
+import {default as NamespaceStore} from 'services/store/store';
+import {parseMetadata} from 'services/metadata-parser';
+import { Dropdown, DropdownMenu, DropdownItem } from 'reactstrap';
+import debounce from 'lodash/debounce';
+import SpotlightModal from 'components/SpotlightSearch/SpotlightModal';
+import Mousetrap from 'mousetrap';
+import T from 'i18n-react';
+require('./SpotlightSearch.less');
+
+const keyMap = {
+  enter: 13,
+  esc: 27,
+  arrowUp: 38,
+  arrowDown: 40
+};
+
+const VIEW_RESULT_LIMIT = 5;
+
+export default class SpotlightSearch extends Component {
+  constructor(props) {
+    super(props);
+
+    this.handleSearchClick = this.handleSearchClick.bind(this);
+    this.debounceSearch = debounce(this.handleSearch.bind(this, event), 300);
+    this.handleCloseSearch = this.handleCloseSearch.bind(this);
+    this.handleToggleModal = this.handleToggleModal.bind(this);
+    this.handleKeyPress = this.handleKeyPress.bind(this);
+
+    this.state = {
+      showSearch: false,
+      showResult: false,
+      showModal: false,
+      searchQuery: '',
+      searchResults: {},
+      focusIndex: 0
+    };
+  }
+
+  componentWillMount() {
+    Mousetrap.bind('?', this.handleSearchClick.bind(this));
+  }
+
+  componentDidUpdate() {
+    if (!this.state.showSearch) { return; }
+    this.spotlightSearch.focus();
+  }
+
+  handleSearchClick(event) {
+    if (event) { event.preventDefault(); }
+
+    this.setState({showSearch: true});
+    let app = document.getElementById('app-container');
+    app.addEventListener('click', this.handleCloseSearch);
+    document.body.addEventListener('keyup', this.handleCloseSearch);
+  }
+
+  handleCloseSearch(event) {
+    // Proceed if and only if:
+    //  1. Event object is null
+    //  2. Escape key was pressed
+    //  3. Target of the event is not the spotlightSearch
+    if (event && event.keyCode !== 27 && event.target === this.spotlightSearch) { return; }
+    let app = document.getElementById('app-container');
+    document.body.removeEventListener('keyup', this.handleCloseSearch);
+    app.removeEventListener('click', this.handleCloseSearch);
+
+    this.setState({showSearch: false, showResult: false});
+  }
+
+  handleResultToggle() {
+    this.setState({
+      showResult: false,
+      showSearch: false
+    });
+  }
+
+  handleSearch() {
+    if (!this.spotlightSearch) { return; }
```

Comment:
On keyboard press of ```?``` it will be null in the beginning, because the spotlight search and keypress event happened at the same cycle